### PR TITLE
Fix infinite loop in nested agg

### LINKED ISSUE #15914: [BUG] Deep level aggregations query hang the request
### Describe the bug

Application in **version 2.16.0 has a bug** which is manifested by hanging while trying to resolve the request - wait for unlimited amount of time. The problem is only for specific query which is included to this bug ticket below.
**Version 1.3.19 handling the query**

### Related component

Search

### To Reproduce

1. Create index:
```
PUT supplier2/
{
  "settings": {
    "number_of_shards": 1,
    "analysis": {
      "analyzer": {
        "default": {
          "tokenizer": "whitespace",
          "filter": [
            "lowercase"
          ]
        }
      },
      "normalizer": {
        "raw_normalizer": {
          "type": "custom",
          "filter": [
            "lowercase"
          ]
        }
      }
    }
  },
  "mappings": {
    "properties": {
      "dataSet": {
        "type": "keyword"
      },
      "supplierValueProperties": {
        "type": "nested",
        "properties": {
          "propertyName": {
            "type": "keyword"
          },
          "propertyType": {
            "type": "keyword"
          },
          "propertyStringValue": {
            "type": "text",
            "fields": {
              "raw": {
                "type": "keyword"
              },
              "raw_normalized": {
                "type": "keyword",
                "normalizer": "raw_normalizer"
              }
            }
          },
          "propertyNumericValue": {
            "type": "double"
          },
          "propertyDateValue": {
            "type": "date",
            "format": "yyyy-MM-dd"
          },
          "propertyBooleanValue": {
            "type": "boolean"
          }
        }
      },
      "references": {
        "type": "nested",
        "properties": {
          "key": {
            "type": "keyword"
          },
          "value": {
            "type": "nested",
            "properties": {
              "id": {
                "type": "keyword"
              },
              "code": {
                "type": "text",
                "fields": {
                  "raw": {
                    "type": "keyword"
                  },
                  "raw_normalized": {
                    "type": "keyword",
                    "normalizer": "raw_normalizer"
                  }
                }
              },
              "referenceValueProperties": {
                "type": "nested",
                "properties": {
                  "propertyName": {
                    "type": "keyword"
                  },
                  "propertyType": {
                    "type": "keyword"
                  },
                  "propertyStringValue": {
                    "type": "text",
                    "fields": {
                      "raw": {
                        "type": "keyword",
                        "ignore_above": 30000
                      },
                      "raw_normalized": {
                        "type": "keyword",
                        "normalizer": "raw_normalizer",
                        "ignore_above": 30000
                      }
                    }
                  },
                  "propertyNumericValue": {
                    "type": "double"
                  },
                  "propertyDateValue": {
                    "type": "date",
                    "format": "yyyy-MM-dd"
                  },
                  "propertyBooleanValue": {
                    "type": "boolean"
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```
2. Execute the query:
```
POST supplier2/_search
{
  "size": 1000,
  "query": {
    "bool": {
      "must": [
        {
          "term": {
            "dataSet": {
              "value": "basic",
              "boost": 1
            }
          }
        }
      ],
      "adjust_pure_negative": true,
      "boost": 1
    }
  },
  "aggregations": {
    "reference_aggregation": {
      "nested": {
        "path": "references"
      },
      "aggregations": {
        "references.key": {
          "terms": {
            "field": "references.key"
          },
          "aggregations": {
            "referenceValueProperties": {
              "nested": {
                "path": "references.value.referenceValueProperties"
              },
              "aggregations": {
                "propertyName": {
                  "terms": {
                    "field": "references.value.referenceValueProperties.propertyName"
                  },
                  "aggregations": {
                    "propertyType": {
                      "terms": {
                        "field": "references.value.referenceValueProperties.propertyType"
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

### Expected behavior

request resolved, like in this version:
  "version" : {
    "distribution" : "opensearch",
    "number" : "1.3.19",
    "build_type" : "zip",
    "build_date" : "2024-08-23T00:39:31.484729800Z",
    "build_snapshot" : false,
    "lucene_version" : "8.10.1",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  }

### Additional Details

default installation, for example in Windows (not related to OS):
"version" : {
    "distribution" : "opensearch",
    "number" : "2.16.0",
    "build_type" : "zip",
    "build_date" : "2024-08-06T20:32:32.086481300Z",
    "build_snapshot" : false,
    "lucene_version" : "9.11.1",
    "minimum_wire_compatibility_version" : "7.10.0",
    "minimum_index_compatibility_version" : "7.0.0"
}

### PR DESCRIPTION
**Pull request by kkewwei**
### Description
Fix infinite loop in nested agg

### Related Issues
Resolves #15914

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -26,6 +26,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Fixed
 - Fix wildcard query containing escaped character ([#15737](https://github.com/opensearch-project/OpenSearch/pull/15737))
 - Fix case-insensitive query on wildcard field ([#15882](https://github.com/opensearch-project/OpenSearch/pull/15882))
+- Fix infinate loop in nested agg ([#15931](https://github.com/opensearch-project/OpenSearch/pull/15931))
```
Comment (dblock) [inline CHANGELOG.md:29]:
```suggestion
- Fix infinite loop in nested agg ([#15931](https://github.com/opensearch-project/OpenSearch/pull/15931))
```

**Review by dblock — CHANGES_REQUESTED** [review]
Typo in CHANGELOG. Sorry!

**Comment by dblock** [comment]
If you are bored, you could add some tests for agg to https://github.com/opensearch-project/opensearch-api-specification and potentially fix any API spec problems with that ;)

**Comment by kkewwei** [comment]
> If you are bored while gradle re-runs, you could add some tests for agg to https://github.com/opensearch-project/opensearch-api-specification and potentially fix any API spec problems with that ;)

Of course, let me have a try.

**Comment by kkewwei** [comment]
> UNSTABLE

org.opensearch.remotestore.multipart.RemoteStoreMultipartIT #11997

**Comment by kkewwei** [comment]
> ❕ Gradle check result for [ab0d430](https://github.com/opensearch-project/OpenSearch/commit/ab0d430adb300135b2c3faca75cc1cbea7596961): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/47841/)
> 
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.remotestore.multipart.RemoteStoreMultipartIT.testNoSearchIdleForAnyReplicaCount
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

org.opensearch.remotestore.multipart.RemoteStoreMultipartIT https://github.com/opensearch-project/OpenSearch/issues/11997

**Review by dblock — APPROVED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/410_nested_aggs.yml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+teardown:
+  - do:
+      indices.delete:
+        index: test_nested_agg_index
+
+---
+"Supported queries":
+  - skip:
+      version: " - 2.99.99"
+      reason: "fixed in 3.0.0"
```
Comment (sandeshkr419) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/410_nested_aggs.yml:48]: I see backport label added to the PR. So most likely, this should get fixed in coming 2.18 release, right?
We should skip this for 2.17.99 ideally then?

**Review by dbwiddis — COMMENTED** [review]
LGTM. I will approve when the checklist at the top of this PR is completed.  Certainly there is testing (box 1) and an API spec (box 2) and it may not need a doc fix (box 3) but I'd prefer to see that assessment done by the PR submitter.

**Comment by reta** [comment]
@kkewwei the issue seems to be not fixed, here is the recent failure (with this pull request merged): https://github.com/opensearch-project/OpenSearch/issues/15914#issuecomment-2361946004

**Comment by kkewwei** [comment]
> @kkewwei the issue seems to be not fixed, here is the recent failure (with this pull request merged): [#15914 (comment)](https://github.com/opensearch-project/OpenSearch/issues/15914#issuecomment-2361946004)

Let me see.

**Pull request merged**